File: kotlin/simple-cipher/src/main/kotlin/SimpleCipher.kt
private val ALPHABET = ('a'..'z').toList()

class Cipher(val key: String = Key.generate()) {
    init {
        require(Key.isValid(key))
    }

    fun encode(input: String) = input.shift(Int::plus)

    fun decode(input: String) = input.shift(Int::minus)

    private fun String.shift(op: (Int, Int) -> Int) =
        mapIndexed { i, c ->
            val alphabetIndex = ALPHABET.indexOf(c)
            val keyIndex = ALPHABET.indexOf(key[i.mod(key.length)])
            ALPHABET[op(alphabetIndex, keyIndex).mod(ALPHABET.size)]
        }.joinToString(separator = "")
}

object Key {
    fun generate() = List(100) { ALPHABET.random() }.joinToString(separator = "")

    fun isValid(key: String) = key.isNotEmpty() && key.all { it in ALPHABET }
}

File: fsharp/simple-cipher/SimpleCipher.fs
module SimpleCipher

open System

let private letters = [| 'a'..'z' |]

module Key =
    let private random = Random()
    let private keyLength = 100
    
    let isValid key = System.Text.RegularExpressions.Regex.IsMatch(key, @"^[a-z]+$")
    
    let generate() = 
        Array.init keyLength (fun _ -> letters.[random.Next(letters.Length)]) 
        |> String

type SimpleCipher(key: string) =
    do 
        if not (Key.isValid key) then 
            invalidArg "key" "Invalid key"
            
    let shift op (input:string) =
        let modulo x y = ((x % y) + y) % y
        let shiftLetter i c =
            let alphabetIndex = Array.IndexOf(letters, c)
            let keyIndex = Array.IndexOf(letters, key.[modulo i key.Length])
            letters.[modulo (op alphabetIndex keyIndex) letters.Length]
        
        String.mapi shiftLetter input
    
    member _.Key with get() = key

    member _.Encode(plaintext: string) = shift (+) plaintext    
    member _.Decode(ciphertext: string) = shift (-) ciphertext
    
    new() = SimpleCipher(Key.generate())
